DataFormats/Common/interface/IdToHitRange.h

```diff
@@ -0,0 +1,220 @@
+        ids_.emplace_back(std::move(id));
+        offsets_.push_back(collection_.size());
+        collection_.insert(collection_.end(), begin, end);
+        if (not isAlreadySorted) {
+          //need to update the indirection table used to find the ranges
+          offsets_.insert(offsets_.begin() + ids_.size() - 1, offsets_.size() - ids_.size());
+        }
+        return;
+      } else if (*i == id) {
+        throw edm::Exception(edm::errors::LogicError, "trying to insert duplicate entry");
+      }
+      if (offsets_.size() == ids_.size()) {
+        //was sorted, but now it will not be so need to reformat offsets_
+        std::vector<unsigned int> newOffsets;
+        newOffsets.reserve(ids_.size() * 2 + 2);
+        for (std::size_t index = 0; index < ids_.size(); ++index) {
+          newOffsets.push_back(index);
+        }
+        for (std::size_t index = 0; index < ids_.size(); ++index) {
+          newOffsets.push_back(offsets_[index]);
+        }
+        offsets_ = std::move(newOffsets);
+      }
+      auto offsetInIds = i - ids_.begin();
+      ids_.insert(i, id);
+      offsets_.push_back(collection_.size());
+      collection_.insert(collection_.end(), begin, end);
+      offsets_.insert(offsets_.begin() + offsetInIds, offsets_.size() - ids_.size());
+    }
+
+    /// perfor post insert action
+    void post_insert() {
+      if (ids_.size() == offsets_.size()) {
+        //already sorted
+        return;
+      }
+
+      std::vector<HIT> sortedCollection;
+      sortedCollection.reserve(collection_.size());
+
+      std::vector<unsigned int> offsets;
+      offsets.reserve(ids_.size());
+
+      for (auto id : ids_) {
+        offsets.push_back(sortedCollection.size());
+        auto range = get(id);
+        sortedCollection.insert(sortedCollection.end(), range.first, range.second);
+      }
+      offsets_ = std::move(offsets);
+      collection_ = std::move(sortedCollection);
+    }
+
+    /// first identifier iterator
+    id_iterator id_begin() const { return ids_.begin(); }
+    /// last identifier iterator
+    id_iterator id_end() const { return ids_.end(); }
+    /// number of contained identifiers
+    size_t id_size() const { return ids_.size(); }
+    /// indentifier vector
+    std::vector<ID> ids() const { return ids_; }
+    /// direct access to an object in the collection
+    reference operator[](size_type i) { return collection_[i]; }
+
+    /// swap member function
+    void swap(IdToHitRange<ID, HIT>& other);
+
+    //Used by ROOT storage
+    CMS_CLASS_VERSION(3)
+
+  private:
+    typename std::vector<ID>::const_iterator find(ID id) const {
+      auto i = std::lower_bound(ids_.begin(), ids_.end(), id);
+      if (i == ids_.end() or *i != id) {
+        return ids_.end();
+      }
+      return i;
+    }
+
+    // ---------- member data --------------------------------
+    std::vector<ID> ids_;
+    std::vector<unsigned int> offsets_;
+    container collection_;
```

Comment:
Could you add some comments how the `ID` element gets mapped to `collection_` via `offsets_`?

Reply:
Is the idea that, previously `map<ID,pair<unsigned int,unsigned int> >` was storing the start and stop indices, and since the container is contiguous (i.e. `start[i+1] = stop[i]`), the offsets is sufficient?

Reply:
yes